text/0000-never-patterns.md

```diff
@@ -0,0 +1,250 @@
+- Feature Name: `never_patterns`
+- Start Date: 2024-10-27
+- RFC PR: [rust-lang/rfcs#3719](https://github.com/rust-lang/rfcs/pull/3719)
+- Rust Issue: [rust-lang/rust#118155](https://github.com/rust-lang/rust/issues/118155)
+
+# Summary
+[summary]: #summary
+
+A `!` pattern indicates a type with no valid values. It is used to indicate an impossible case when matching an empty type in unsafe code.
+
+```rust
+enum Void {}
+unsafe fn empty_tup<T>(tup: *const (T, Void)) -> ! {
+    unsafe {
+        match *tup { ! }
+    }
+}
+```
+
+Note: this RFC is purely about a new pattern syntax. It does not propose changes to exhaustiveness checking or to the operational semantics of uninhabited types.
+
+
+# Motivation
+[motivation]: #motivation
+
+Rust's unsafe semantics are access-based: the validity of data only matters when reading from/writing to a place. When pattern-matching, patterns only access the data (discriminants/values) they need to choose the right arm.
+
+Empty types are funny because the natural way of matching them is to write no arms at all:
+
+```rust
+fn empty_tup<T>(tup: (T, !)) -> ! {
+    match tup {}
+}
+```
+
+Here the absence of an arm plays the role of a "read discriminant" kind of operation. This is fine in this case, but around unsafe code that interacts with possibly-uninitialized data, accesses should be explicitly visible in the pattern.
+
+Today, when matching empty types inside places that may contain uninitialized data, rust requires a dummy match arm:
+```rust
+enum Void {}
+unsafe fn empty_tup<T>(tup: *const (T, Void)) -> ! {
+    unsafe {
+        match *tup {
+            _ => unreachable_unchecked(),
+            // or
+            (_, x) => match x {}
```

Comment:
```suggestion
            (_, x) => match x {}
```
The first example seems odd and orthogonal to the RFC.

Reply:
Maybe `_ => unreachable!()` is better? That's just what people commonly write in this case

Reply:
It may be what people commonly write, but the intent of this PR is better described by `match x {}` so IMO we should have only that.